packages/worker/src/worker.ts

```diff
@@ -821,21 +918,94 @@ export class Worker {
    * Errors from core responses are translated to cancellation requests and fed back via the activityFeedbackSubject.
    */
   protected activityHeartbeat$(): Observable<void> {
+    function process(state: HeartbeatState, heartbeat: Heartbeat): HeartbeatStateAndOutput {
+      return { state: { ...state, processing: true, pending: undefined }, output: { type: 'send', heartbeat } };
+    }
+
+    function storePending(state: HeartbeatState, heartbeat: Heartbeat): HeartbeatStateAndOutput {
+      return { state: { ...state, pending: heartbeat }, output: null };
+    }
+
+    function complete(callback?: () => void): HeartbeatStateAndOutput {
+      return {
+        state: { pending: undefined, completionCallback: undefined, processing: false, closed: true },
+        output: { type: 'close', completionCallback: callback },
+      };
+    }
+
     return this.activityHeartbeatSubject.pipe(
       // The only way for this observable to be closed is by state changing to DRAINED meaning that all in-flight activities have been resolved and thus there should not be any heartbeats to send.
       this.takeUntilState('DRAINED'),
       tap({
-        next: ({ taskToken }) => this.log.trace('Got activity heartbeat', { taskToken: formatTaskToken(taskToken) }),
+        next: ({ base64TaskToken }) => this.log.trace('Got activity heartbeat', { taskToken: base64TaskToken }),
         complete: () => this.log.debug('Heartbeats complete'),
       }),
-      mergeMap(async ({ taskToken, details }) => {
-        const payload = await encodeToPayload(this.options.loadedDataConverter, details);
-        const arr = coresdk.ActivityHeartbeat.encodeDelimited({
-          taskToken,
-          details: [payload],
-        }).finish();
-        this.nativeWorker.recordActivityHeartbeat(byteArrayToBuffer(arr));
-      })
+      closeableGroupBy(({ base64TaskToken }) => base64TaskToken),
+      mergeMap((group$) =>
+        group$.pipe(
+          mapWithState(
+            (state: HeartbeatState, input: HeartbeatInput): HeartbeatStateAndOutput => {
+              // Ignore any input if we've marked this activity heartbeat stream as closed
+              if (state.closed) return { state, output: null };
+
+              switch (input.type) {
```

Comment:
All this extra logic is a bit unfortunate. Now you've basically done the same amount of work Core had to do to ensure they get flushed. :cry: 

I don't know that there's any way around it with the async converters though...

Reply:
Yeah, that's unfortunate..